### DISCUSSION TITLE
Passing a macro in rendering context so that the macro can be called dynamically

### DISCUSSION THREAD

**Discussion by CaliViking** [mitsuhiko/minijinja#463] [Q&A] [answered]
Is it possible to pass the name of a macro into the context while rendering and then allow the template to use the passed macro?

I am looking for a dynamic way to render multiple views of the same data. I am implementing the views as macros as they can be called with parameters.

I can currently do this by creating a template for every view and then calling the macro in that template. I would like to avoid having to create both a macro and then the template.

It looks like I can call a function dynamically, but I have found no way to call a macro dynamically

If this does not work, then I will likely implement everything in a template. It is not the end of the world, but it is slightly less reusable.

Thanks in advance!

**Comment by mitsuhiko** [comment, marked as answer]
Unfortunately passing actual macros to another template is not supported today due to technical limitations: #329

As a workaround you could import the macros into a meta template and then call them by name:

`all-macros.html`:

```jinja
{% from "./macros1.html" import macro1 %}
{% from "./macros2.html" import macro2 %}
```

example.html:

```jinja
{% import "all-macros.html as macros %}
{{ macros[macro_name]() }}
```

If the variable `macro_name` is then `macro2` it will invoke `macro2`.

**Reply by CaliViking** [reply]
@mitsuhiko,
Thank you so much for a quick response over the easter holiday weekend.
This is an interesting approach. I will try it.

Another approach that may be easier to implement than passing macros around is if we could access, call, and render macros directly from Rust, similar to how we can render a template.

**Reply by CaliViking** [reply]
@mitsuhiko ,

I just tried it and it works. Thank you.

What is the performance impact of calling a macro with this approach compared to having the text inline in the template? I cannot quite nail it down, but it looks like there could be an impact that is somewhere in the 0.1 mSec range per call when I run in debug mode.

Edit: I just tested this on a larger scale in release mode and I cannot see a significant performance difference.
